File: algorithm/knapsack-dp.js
/*
아주 작은 sub problem 풀어서 큰 문제를 푸는 것! --> 추상적

작은 것 부터 시작하지만(선택할 수도 A, 안할수도 B) 내가 A라는걸 선택 할 수 있고 B라는 걸 선택 할 수 있다.

item과 cap(수용량)을 중요하게 봐라!

base case ---> input = [] or cap = 0

        
    dp table

        0 1 2 3 4 5 6 7 8 9 10
        
   []   0 0 0 0 0 0 0 0 0 0 0
[1,2]   0 0 1 1 1 1 1 1 1 1 1
[4,3]   0 0 1 4 4 5             ---> formulation 
 ^ ^
 value,weight                                                 현재 벨류,
                    dp[][]             dp[i][c]= math.max(dp[i-1][c], dp[i-1][c-curweight]+culVal)
                      ^ ^
                 [1,2],[4,3]  index, cab 
                   ^     ^
           index   0     1
!!중요!!! 패턴 찾기!!!


*/

File: algorithm/knapsack-dp.py
#### DP
"""
itmes =  [
    [465,100],
    [400,85],
    [255,55],
    [350,45],
    [650,130],
    [1000,190],
    [455,100],
    [100,25],
    [1200,190],
    [320,65],
    [750,100],
    [50,45],
    [550,65],
    [100,50],
    [600,70],
    [255,40]
]
items의 첫번째 값은 value, 두번재는 weight
cap = 200
res = [1505,[7,12,14,15]]


            items   cap
base case -> []  , 0

dpTable

cap             0 1 2 3 4 5 6 7 8 9 10
index 0 []      0 0 0 0 0 0 0 0 0 0 0
      1 [1,2]   0 0 1 1 1 1 1 1 1 1 1
      2 [4,3]   0 0 1 4 4 5 5 5 5 5 5   
      3 [5,6]   0 0 1 4 4 5 5 5 6 9 9   
      4 [6,7]   0 0 1 4 4 5 5 6 6 9 10  
                                             
cap = 10

  dp[i][c] = max(dp[i-1][c],dp[i-1][c-curW]+curV)
  
  output = [10,[1,3]]



"""

itmes =  [
    [465,100],
    [400,85],
    [255,55],
    [350,45],
    [650,130],
    [1000,190],
    [455,100],
    [100,25],
    [1200,190],
    [320,65],
    [750,100],
    [50,45],
    [550,65],
    [100,50],
    [600,70],
    [255,40]
]

cap = 200

dp = []

# init dpTable
for i in range(0,len(itmes)+1):
    dp.append([])
    for j in range(0,cap+1):
        dp[i].append(0)

# apply formula
for i in range(1,len(itmes)+1):
    curValue = itmes[i-1][0]
    curWeight = itmes[i-1][1]
    for c in range(1,cap+1):
        if c < curWeight:
            dp[i][c] = dp[i-1][c]
        else:
            dp[i][c] = max(dp[i-1][c], dp[i-1][c-curWeight]+curValue)


curIndex = len(itmes)
w  = cap


print(dp[len(itmes)][cap])
